presto-kafka/src/main/java/io/prestosql/plugin/kafka/KafkaFilterManager.java

```diff
@@ -0,0 +1,243 @@
+    {
+        Long low = INVALID_KAFKA_RANGE_INDEX;
+        Long high = INVALID_KAFKA_RANGE_INDEX;
+        if (domain.isSingleValue()) {
+            low = (long) domain.getSingleValue();
+            high = (long) domain.getSingleValue();
+        }
+        else {
+            ValueSet valueSet = domain.getValues();
+            if (valueSet instanceof SortedRangeSet) {
+                Ranges ranges = ((SortedRangeSet) valueSet).getRanges();
+                List<io.prestosql.spi.predicate.Range> rangeList = ranges.getOrderedRanges();
+                if (rangeList.stream().allMatch(io.prestosql.spi.predicate.Range::isSingleValue)) {
+                    List<Long> values = rangeList.stream()
+                            .map(range -> (Long) range.getSingleValue())
+                            .collect(toImmutableList());
+                    low = Collections.min(values);
+                    high = Collections.max(values);
+                }
+                else {
+                    Marker lowMark = ranges.getSpan().getLow();
+                    low = getLowByLowMark(lowMark).orElse(low);
+                    Marker highMark = ranges.getSpan().getHigh();
+                    high = getHighByHighMark(highMark).orElse(high);
+                }
+            }
+        }
+        if (high != INVALID_KAFKA_RANGE_INDEX) {
+            high = high + 1;
+        }
+        return Optional.of(new Range(low, high));
+    }
+
+    private static Set<Long> getSingleSet(Domain domain, Set<Long> sourceValues)
+    {
+        requireNonNull(sourceValues, "sourceValues is none");
+        if (domain.isSingleValue()) {
+            long singleValue = (long) domain.getSingleValue();
+            return sourceValues.stream().filter(sourceValue -> sourceValue == singleValue).collect(toImmutableSet());
+        }
+        else {
+            ValueSet valueSet = domain.getValues();
+            if (valueSet instanceof SortedRangeSet) {
+                Ranges ranges = ((SortedRangeSet) valueSet).getRanges();
+                List<io.prestosql.spi.predicate.Range> rangeList = ranges.getOrderedRanges();
+                if (rangeList.stream().allMatch(io.prestosql.spi.predicate.Range::isSingleValue)) {
+                    return rangeList.stream()
+                            .map(range -> (Long) range.getSingleValue())
+                            .filter(sourceValues::contains)
+                            .collect(toImmutableSet());
+                }
+                else {
+                    long low = sourceValues.stream().min(Long::compareTo).orElse(0L);
+                    long high = sourceValues.stream().max(Long::compareTo).orElse(0L);
+
+                    Marker lowMark = ranges.getSpan().getLow();
+                    Optional<Long> lowByLowMark = getLowByLowMark(lowMark);
+                    if (lowByLowMark.isPresent()) {
+                        low = Long.max(lowByLowMark.get(), low);
```

Comment:
Maybe define helper method `long min(long a, Optional<Marker> b)` and `max ....` and use it here and in `getRange` above

Reply:
Thanks. will resolve that.